File: src/lib/fileBridge.ts
/**
 * WebSocket File Bridge Client
 *
 * Connects to the backend WebSocket endpoint for real-time file sync
 * between IndexedDB (persistent, source of truth) and the sandbox (ephemeral).
 *
 * Replaces the broken KV-as-bridge pattern:
 *   Before: IDB ←→ KV (non-persistent) ←→ Sandbox
 *   After:  IDB ←→ WebSocket Bridge ←→ Sandbox
 *
 * Protocol:
 *   Frontend → Server:
 *     {type: "sync_all", files: Record<string, string>}
 *     {type: "file_write", path: string, content: string}
 *     {type: "file_delete", path: string}
 *     {type: "file_list_request"}
 *     {type: "ping"}
 *
 *   Server → Frontend:
 *     {type: "connected", conversation_id: string}
 *     {type: "file_write", path: string, content: string}
 *     {type: "file_delete", path: string}
 *     {type: "sync_all", files: Record<string, string>, count: number}
 *     {type: "sync_ack", count: number}
 *     {type: "file_list_response", files: Record<string, string>}
 *     {type: "pong"}
 *     {type: "error", message: string}
 */

import { saveFile, deleteFile, loadConversationFiles } from './workspaceStore';

const RECONNECT_DELAY = 2000;
const MAX_RECONNECT_DELAY = 30000;
const PING_INTERVAL = 25000;

export interface FileBridgeCallbacks {
  onFileWrite?: (path: string, content: string) => void;
  onFileDelete?: (path: string) => void;
  onSyncAll?: (files: Record<string, string>) => void;
  onConnected?: () => void;
  onDisconnected?: () => void;
  onError?: (message: string) => void;
}

export class FileBridgeClient {
  private ws: WebSocket | null = null;
  private conversationId: string = '';
  private callbacks: FileBridgeCallbacks = {};
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private reconnectDelay: number = RECONNECT_DELAY;
  private intentionalClose: boolean = false;
  private connected: boolean = false;

  /**
   * Connect to the file bridge WebSocket.
   */
  connect(conversationId: string, callbacks: FileBridgeCallbacks = {}): void {
    this.conversationId = conversationId;
    this.callbacks = callbacks;
    this.intentionalClose = false;
    this._doConnect();
  }

  /**
   * Disconnect from the file bridge.
   */
  disconnect(): void {
    this.intentionalClose = true;
    this._cleanup();
  }

  /**
   * Send full IDB state to server (on connect or reconnect).
   */
  async syncToServer(conversationId?: string): Promise<void> {
    const cid = conversationId || this.conversationId;
    if (!cid) return;

    try {
      const files = await loadConversationFiles(cid);
      if (Object.keys(files).length > 0) {
        this._send({ type: 'sync_all', files });
      }
    } catch (err) {
      console.warn('[file-bridge] Failed to sync files to server:', err);
    }
  }

  /**
   * Push a file write to the server.
   */
  pushFileWrite(path: string, content: string): void {
    this._send({ type: 'file_write', path, content });
  }

  /**
   * Push a file deletion to the server.
   */
  pushFileDelete(path: string): void {
    this._send({ type: 'file_delete', path });
  }

  /**
   * Request current file list from server.
   */
  requestFileList(): void {
    this._send({ type: 'file_list_request' });
  }

  /**
   * Check if currently connected.
   */
  isConnected(): boolean {
    return this.connected;
  }

  // ── Private ────────────────────────────────────────────────────────

  private _doConnect(): void {
    if (this.ws) {
      this._cleanup();
    }

    const proto = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
    const url = `${proto}//${window.location.host}/ws/file-bridge`;

    try {
      this.ws = new WebSocket(url);
    } catch (err) {
      console.warn('[file-bridge] WebSocket creation failed:', err);
      this._scheduleReconnect();
      return;
    }

    this.ws.onopen = () => {
      console.log('[file-bridge] Connected');
      this.connected = true;
      this.reconnectDelay = RECONNECT_DELAY;
      this._startPing();
      this.callbacks.onConnected?.();

      // Send initial IDB state to server
      this.syncToServer();
    };

    this.ws.onmessage = (event) => {
      this._handleMessage(event.data);
    };

    this.ws.onerror = (event) => {
      console.warn('[file-bridge] Error:', event);
      this.callbacks.onError?.('WebSocket error');
    };

    this.ws.onclose = () => {
      console.log('[file-bridge] Disconnected');
      this.connected = false;
      this._stopPing();
      this.callbacks.onDisconnected?.();

      if (!this.intentionalClose) {
        this._scheduleReconnect();
      }
    };
  }

  private _handleMessage(data: string): void {
    let msg: any;
    try {
      msg = JSON.parse(data);
    } catch {
      console.warn('[file-bridge] Invalid JSON:', data);
      return;
    }

    const cid = this.conversationId;

    switch (msg.type) {
      case 'connected':
        console.log(`[file-bridge] Server acknowledged: ${msg.conversation_id}`);
        break;

      case 'file_write':
        // Server pushed a file write from sandbox → save to IDB
        if (msg.path && typeof msg.content === 'string') {
          saveFile({ conversationId: cid, filepath: msg.path, content: msg.content })
            .catch(err => console.warn('[file-bridge] IDB save failed:', err));
          this.callbacks.onFileWrite?.(msg.path, msg.content);
        }
        break;

      case 'file_delete':
        // Server pushed a file deletion from sandbox → delete from IDB
        if (msg.path) {
          deleteFile(cid, msg.path)
            .catch(err => console.warn('[file-bridge] IDB delete failed:', err));
          this.callbacks.onFileDelete?.(msg.path);
        }
        break;

      case 'sync_all':
        // Server sent full file state → replace IDB
        if (msg.files && typeof msg.files === 'object') {
          const files = msg.files as Record<string, string>;
          // Bulk save to IDB
          (async () => {
            for (const [path, content] of Object.entries(files)) {
              await saveFile({ conversationId: cid, filepath: path, content });
            }
          })().catch(err => console.warn('[file-bridge] Bulk IDB save failed:', err));
          this.callbacks.onSyncAll?.(files);
        }
        break;

      case 'sync_ack':
        console.log(`[file-bridge] Server acknowledged sync: ${msg.count} files`);
        break;

      case 'file_list_response':
        if (msg.files && typeof msg.files === 'object') {
          this.callbacks.onSyncAll?.(msg.files);
        }
        break;

      case 'pong':
        // Heartbeat response
        break;

      case 'error':
        console.warn(`[file-bridge] Server error: ${msg.message}`);
        this.callbacks.onError?.(msg.message);
        break;

      default:
        console.warn(`[file-bridge] Unknown message type: ${msg.type}`);
    }
  }

  private _send(msg: object): void {
    if (this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(msg));
    }
  }

  private _startPing(): void {
    this._stopPing();
    this.pingTimer = setInterval(() => {
      this._send({ type: 'ping' });
    }, PING_INTERVAL);
  }

  private _stopPing(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
  }

  private _scheduleReconnect(): void {
    if (this.reconnectTimer) return;
    if (this.intentionalClose) return;

    console.log(`[file-bridge] Reconnecting in ${this.reconnectDelay}ms...`);
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this._doConnect();
    }, this.reconnectDelay);

    // Exponential backoff
    this.reconnectDelay = Math.min(this.reconnectDelay * 2, MAX_RECONNECT_DELAY);
  }

  private _cleanup(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    this._stopPing();
    if (this.ws) {
      this.ws.onopen = null;
      this.ws.onmessage = null;
      this.ws.onerror = null;
      this.ws.onclose = null;
      this.ws.close();
      this.ws = null;
    }
    this.connected = false;
  }
}

// Singleton instance
export const fileBridge = new FileBridgeClient();

File: agents/chat/file_bridge.py
"""
WebSocket File Bridge — Persistent IDB ↔ Ephemeral Sandbox sync.

Replaces the broken KV-as-bridge pattern with a direct WebSocket connection
between the frontend (IndexedDB = source of truth) and the sandbox container.

Protocol:
  Frontend → Sandbox:
    {"type": "file_write", "path": "...", "content": "..."}
    {"type": "file_delete", "path": "..."}
    {"type": "file_list_request"}
    {"type": "sync_all", "files": {"path": "content", ...}}

  Sandbox → Frontend:
    {"type": "file_write", "path": "...", "content": "..."}
    {"type": "file_delete", "path": "..."}
    {"type": "file_list_response", "files": {"path": "content", ...}}
    {"type": "sync_ack", "count": N}
    {"type": "error", "message": "..."}

Flow:
  1. Frontend connects via WebSocket after auth
  2. Frontend sends full IDB state as initial sync_all
  3. Sandbox tools write locally → push changes back via file_write
  4. Frontend receives changes → saves to IDB → updates sidebar
  5. On reconnect, frontend re-syncs full IDB state
"""

from __future__ import annotations

import asyncio
import json
from typing import Any

from .._logger import create_logger

logger = create_logger("file_bridge")


class FileBridge:
    """Manages WebSocket connections for file sync between frontend and sandbox."""

    def __init__(self) -> None:
        self._connections: dict[str, Any] = {}  # cid -> websocket
        self._file_caches: dict[str, dict[str, str]] = {}  # cid -> files

    def register(self, cid: str, ws: Any) -> None:
        """Register a WebSocket connection for a conversation."""
        self._connections[cid] = ws
        if cid not in self._file_caches:
            self._file_caches[cid] = {}
        logger.log(f"[bridge] Registered connection for {cid}")

    def unregister(self, cid: str) -> None:
        """Unregister a WebSocket connection."""
        self._connections.pop(cid, None)
        logger.log(f"[bridge] Unregistered connection for {cid}")

    def is_connected(self, cid: str) -> bool:
        """Check if a conversation has an active WebSocket connection."""
        return cid in self._connections

    async def send_to_frontend(self, cid: str, message: dict[str, Any]) -> bool:
        """Send a message to the frontend via WebSocket."""
        ws = self._connections.get(cid)
        if not ws:
            return False
        try:
            await ws.send(json.dumps(message))
            return True
        except Exception as e:
            logger.log(f"[bridge] Failed to send to {cid}: {e}")
            self.unregister(cid)
            return False

    async def push_file_write(self, cid: str, path: str, content: str) -> bool:
        """Push a file write from sandbox to frontend (for IDB persistence)."""
        # Update cache
        if cid in self._file_caches:
            self._file_caches[cid][path] = content
        return await self.send_to_frontend(cid, {
            "type": "file_write",
            "path": path,
            "content": content,
        })

    async def push_file_delete(self, cid: str, path: str) -> bool:
        """Push a file deletion from sandbox to frontend."""
        if cid in self._file_caches:
            self._file_caches[cid].pop(path, None)
        return await self.send_to_frontend(cid, {
            "type": "file_delete",
            "path": path,
        })

    async def push_full_sync(self, cid: str, files: dict[str, str]) -> bool:
        """Push full file state to frontend."""
        if cid in self._file_caches:
            self._file_caches[cid] = dict(files)
        return await self.send_to_frontend(cid, {
            "type": "sync_all",
            "files": files,
            "count": len(files),
        })

    def get_cached_files(self, cid: str) -> dict[str, str]:
        """Get cached files for a conversation."""
        return dict(self._file_caches.get(cid, {}))

    def update_cache(self, cid: str, files: dict[str, str]) -> None:
        """Update the file cache from frontend sync."""
        if cid not in self._file_caches:
            self._file_caches[cid] = {}
        self._file_caches[cid].update(files)


# Singleton instance
_bridge = FileBridge()


def get_bridge() -> FileBridge:
    """Get the global FileBridge instance."""
    return _bridge


async def handle_websocket(context: Any) -> None:
    """Handle WebSocket connection for file bridge.

    This is the EdgeOne entry point for the /ws/file-bridge route.
    """
    cid = context.conversation_id
    ws = context.websocket

    bridge = get_bridge()
    bridge.register(cid, ws)

    try:
        # Send initial acknowledgment
        await ws.send(json.dumps({
            "type": "connected",
            "conversation_id": cid,
        }))

        # Listen for messages from frontend
        async for raw_message in ws:
            try:
                msg = json.loads(raw_message)
                msg_type = msg.get("type")

                if msg_type == "sync_all":
                    # Frontend sends full IDB state
                    files = msg.get("files", {})
                    bridge.update_cache(cid, files)
                    logger.log(f"[bridge] Received sync_all from {cid}: {len(files)} files")
                    await ws.send(json.dumps({
                        "type": "sync_ack",
                        "count": len(files),
                    }))

                elif msg_type == "file_write":
                    # Frontend sends a single file update
                    path = msg.get("path", "")
                    content = msg.get("content", "")
                    if path:
                        bridge.update_cache(cid, {path: content})
                        logger.log(f"[bridge] Received file_write from {cid}: {path}")

                elif msg_type == "file_delete":
                    # Frontend deletes a file
                    path = msg.get("path", "")
                    if path and cid in bridge._file_caches:
                        bridge._file_caches[cid].pop(path, None)
                        logger.log(f"[bridge] Received file_delete from {cid}: {path}")

                elif msg_type == "file_list_request":
                    # Frontend requests current file list
                    files = bridge.get_cached_files(cid)
                    await ws.send(json.dumps({
                        "type": "file_list_response",
                        "files": files,
                    }))

                elif msg_type == "ping":
                    await ws.send(json.dumps({"type": "pong"}))

                else:
                    await ws.send(json.dumps({
                        "type": "error",
                        "message": f"Unknown message type: {msg_type}",
                    }))

            except json.JSONDecodeError:
                await ws.send(json.dumps({
                    "type": "error",
                    "message": "Invalid JSON",
                }))
            except Exception as e:
                logger.log(f"[bridge] Error processing message: {e}")

    except Exception as e:
        logger.log(f"[bridge] Connection error for {cid}: {e}")
    finally:
        bridge.unregister(cid)
